lib/IRGen/GenCall.cpp

```diff
@@ -667,13 +664,26 @@ void SignatureExpansion::expandCoroutineResult(bool forContinuation) {
       llvm_unreachable("should have been filtered out before here");
 
     // Yield-once coroutines just return void from the continuation.
-    case SILCoroutineKind::YieldOnce:
-      ResultIRType = IGM.VoidTy;
+    case SILCoroutineKind::YieldOnce: {
+      auto fnConv = getSILFuncConventions();
+
+      assert(fnConv.getNumIndirectSILResults() == 0);
+      // Ensure that no parameters were added before to correctly record their ABI
+      // details.
+      assert(ParamIRTypes.empty());
+
+      // Expand the direct result.
+      const TypeInfo *directResultTypeInfo;
+      std::tie(ResultIRType, directResultTypeInfo) = expandDirectResult();
+
       return;
+    }
 
     // Yield-many coroutines yield the same types from the continuation
     // as they do from the ramp function.
     case SILCoroutineKind::YieldMany:
+      assert(FnType->getNumResults() == 0 &&
+             "having both normal and yield results is currently unsupported");
```

Comment:
I don't know why this would be different from the yield-once case.

Reply:
The lowering to LLVM IR is different for yield-once and yield-many coroutines. The latter return null coroutine frame pointer to indicate the completion. As there is no usage examples for them (plus the results) I decided not to proceed with allowing results for them. So only yield-once coroutines might have normal results.

Reply:
Okay, fair enough.